### ISSUE TITLE
hiredis-test hangs when -DNDEBUG

### ISSUE THREAD

**Issue by keith-bennett-gbg** [redis/hiredis#308]
In a cmake project, I'm including hiredis as an external custom target. When the `CMAKE_BUILD_TYPE` is a release-type build (`RelWithDebInfo`, `Release`, etc), cmake will typically pass `-DNDEBUG` as a compiler flag to the external `make`.

This seemed to have caused `hiredis-test` to hang in some tests. It appears that `test.c` is using `assert()` with statements which have side effects, which is actually [a fairly big no-no](http://stackoverflow.com/questions/10593492/catching-assert-with-side-effects).

**Comment by badboy** [comment]
Side-effects in assert in the tests is one thing (they are tests afterall), but we have atleast one place relying on a function call wrapped in `assert`: https://github.com/redis/hiredis/blob/master/async.c#L339

**Comment by mattsta** [comment]
Yeah, it's not good, but it's there for now.  It's a Redis style thing too.  assert() gets used to check the return values of functions with side effects and it makes using code in other projects tricky/annoying because you have to manually replace all the asserts.

Status: submit PRs to fix things or what until we get around to fixing it on our own schedule.  :)

**Comment by badboy** [comment]
Redis doesn't need to interact with other code bases, hiredis does.
Do we want the explicit check to be there always or would something like this be enough for now:

```diff
diff --git i/async.c w/async.c
index 84480bc..9a494c2 100644
--- i/async.c
+++ w/async.c
@@ -336,7 +336,8 @@ static void __redisAsyncDisconnect(redisAsyncContext *ac) {
 
     if (ac->err == 0) {
         /* For clean disconnects, there should be no pending callbacks. */
-        assert(__redisShiftCallback(&ac->replies,NULL) == REDIS_ERR);
+        int err = __redisShiftCallback(&ac->replies,NULL);
+        assert(err == REDIS_ERR);
     } else {
         /* Disconnection is caused by an error, make sure that pending
          * callbacks cannot call new commands. */
```

**Comment by mattsta** [comment]
oh, I didn't look at how many places it happens.  Just that once?  Not too bad then.

The diff above is _only_ debugging, so it's safe to keep in the assert.  Looks like just the asserts in the tests are the problem here.  But, they are used only for testing, so, it still seems valid.

**Comment by keith-bennett-gbg** [comment]
Strictly speaking, it would change the signature since not calling `__redisShiftCallback()` would end up not `free()`ing `ac->replies->cb` and incrementing the linked list to the next head. However, the comment clearly indicates that should be `NULL` anyway.

**Comment by mattsta** [comment]
> However, the comment clearly indicates that should be NULL anyway.

Yeah, if it *does* free something, it would return REDIS_OK, then your program would crash at the assert.  The assert in that case is just verifying other data structures are already empty, and if they aren't, the program is wrong and you crash.

**Comment by badboy** [comment]
No other function-calls in assert in the code except in `test.c`

**Comment by keith-bennett-gbg** [comment]
Well it won't let me push a branch to create a pull request. So here's a proposed diff

```diff
From 7b5f1057f47196c1da71eafba111fa3ff748f6f3 Mon Sep 17 00:00:00 2001
From: Keithb <keithb@genebygene.com>
Date: Mon, 16 Feb 2015 09:08:46 -0600
Subject: [PATCH] * Fix tests when assert() is no-op (such as when -DNDEBUG, eg
 when included in cmake as an external project) * Resolves issue #308

---
 test.c | 44 +++++++++++++++++++++++++-------------------
 1 file changed, 25 insertions(+), 19 deletions(-)

diff --git a/test.c b/test.c
index f1566fb..a1a31c7 100644
--- a/test.c
+++ b/test.c
@@ -43,17 +43,23 @@ static long long usec(void) {
     return (((long long)tv.tv_sec)*1000000)+tv.tv_usec;
 }
 
+#ifdef NDEBUG
+#define ASSERT(statement) (void)(statement)
+#else
+#define ASSERT(statement) assert(statement)
+#endif
+
 static redisContext *select_database(redisContext *c) {
     redisReply *reply;
 
     /* Switch to DB 9 for testing, now that we know we can chat. */
     reply = redisCommand(c,"SELECT 9");
-    assert(reply != NULL);
+    ASSERT(reply != NULL);
     freeReplyObject(reply);
 
     /* Make sure the DB is emtpy */
     reply = redisCommand(c,"DBSIZE");
-    assert(reply != NULL);
+    ASSERT(reply != NULL);
     if (reply->type == REDIS_REPLY_INTEGER && reply->integer == 0) {
         /* Awesome, DB 9 is empty and we can continue. */
         freeReplyObject(reply);
@@ -70,10 +76,10 @@ static int disconnect(redisContext *c, int keep_fd) {
 
     /* Make sure we're on DB 9. */
     reply = redisCommand(c,"SELECT 9");
-    assert(reply != NULL);
+    ASSERT(reply != NULL);
     freeReplyObject(reply);
     reply = redisCommand(c,"FLUSHDB");
-    assert(reply != NULL);
+    ASSERT(reply != NULL);
     freeReplyObject(reply);
 
     /* Free the context as well, but keep the fd if requested. */
@@ -99,7 +105,7 @@ static redisContext *connect(struct config config) {
             c = redisConnectFd(fd);
         }
     } else {
-        assert(NULL);
+        ASSERT(NULL);
     }
 
     if (c == NULL) {
@@ -232,7 +238,7 @@ static void test_append_formatted_commands(struct config config) {
 
     test_cond(redisAppendFormattedCommand(c, cmd, len) == REDIS_OK);
 
-    assert(redisGetReply(c, (void*)&reply) == REDIS_OK);
+    ASSERT(redisGetReply(c, (void*)&reply) == REDIS_OK);
 
     free(cmd);
     freeReplyObject(reply);
@@ -291,7 +297,7 @@ static void test_reply_reader(void) {
     reader->fn = NULL;
     redisReaderFeed(reader,(char*)"+OK\r",4);
     ret = redisReaderGetReply(reader,&reply);
-    assert(ret == REDIS_OK && reply == NULL);
+    ASSERT(ret == REDIS_OK && reply == NULL);
     redisReaderFeed(reader,(char*)"\n",1);
     ret = redisReaderGetReply(reader,&reply);
     test_cond(ret == REDIS_OK && reply == (void*)REDIS_REPLY_STATUS);
@@ -302,7 +308,7 @@ static void test_reply_reader(void) {
     reader->fn = NULL;
     redisReaderFeed(reader,(char*)"x",1);
     ret = redisReaderGetReply(reader,&reply);
-    assert(ret == REDIS_ERR);
+    ASSERT(ret == REDIS_ERR);
     ret = redisReaderGetReply(reader,&reply);
     test_cond(ret == REDIS_ERR && reply == NULL);
     redisReaderFree(reader);
@@ -474,14 +480,14 @@ static void test_blocking_io_errors(struct config config) {
      * On >2.0, QUIT will return with OK and another read(2) needed to be
      * issued to find out the socket was closed by the server. In both
      * conditions, the error will be set to EOF. */
-    assert(c->err == REDIS_ERR_EOF &&
+    ASSERT(c->err == REDIS_ERR_EOF &&
         strcmp(c->errstr,"Server closed the connection") == 0);
     redisFree(c);
 
     c = connect(config);
     test("Returns I/O error on socket timeout: ");
     struct timeval tv = { 0, 1000 };
-    assert(redisSetTimeout(c,tv) == REDIS_OK);
+    ASSERT(redisSetTimeout(c,tv) == REDIS_OK);
     test_cond(redisGetReply(c,&_reply) == REDIS_ERR &&
         c->err == REDIS_ERR_IO && errno == EAGAIN);
     redisFree(c);
@@ -526,7 +532,7 @@ static void test_throughput(struct config config) {
     t1 = usec();
     for (i = 0; i < num; i++) {
         replies[i] = redisCommand(c,"PING");
-        assert(replies[i] != NULL && replies[i]->type == REDIS_REPLY_STATUS);
+        ASSERT(replies[i] != NULL && replies[i]->type == REDIS_REPLY_STATUS);
     }
     t2 = usec();
     for (i = 0; i < num; i++) freeReplyObject(replies[i]);
@@ -537,8 +543,8 @@ static void test_throughput(struct config config) {
     t1 = usec();
     for (i = 0; i < num; i++) {
         replies[i] = redisCommand(c,"LRANGE mylist 0 499");
-        assert(replies[i] != NULL && replies[i]->type == REDIS_REPLY_ARRAY);
-        assert(replies[i] != NULL && replies[i]->elements == 500);
+        ASSERT(replies[i] != NULL && replies[i]->type == REDIS_REPLY_ARRAY);
+        ASSERT(replies[i] != NULL && replies[i]->elements == 500);
     }
     t2 = usec();
     for (i = 0; i < num; i++) freeReplyObject(replies[i]);
@@ -551,8 +557,8 @@ static void test_throughput(struct config config) {
         redisAppendCommand(c,"PING");
     t1 = usec();
     for (i = 0; i < num; i++) {
-        assert(redisGetReply(c, (void*)&replies[i]) == REDIS_OK);
-        assert(replies[i] != NULL && replies[i]->type == REDIS_REPLY_STATUS);
+        ASSERT(redisGetReply(c, (void*)&replies[i]) == REDIS_OK);
+        ASSERT(replies[i] != NULL && replies[i]->type == REDIS_REPLY_STATUS);
     }
     t2 = usec();
     for (i = 0; i < num; i++) freeReplyObject(replies[i]);
@@ -564,9 +570,9 @@ static void test_throughput(struct config config) {
         redisAppendCommand(c,"LRANGE mylist 0 499");
     t1 = usec();
     for (i = 0; i < num; i++) {
-        assert(redisGetReply(c, (void*)&replies[i]) == REDIS_OK);
-        assert(replies[i] != NULL && replies[i]->type == REDIS_REPLY_ARRAY);
-        assert(replies[i] != NULL && replies[i]->elements == 500);
+        ASSERT(redisGetReply(c, (void*)&replies[i]) == REDIS_OK);
+        ASSERT(replies[i] != NULL && replies[i]->type == REDIS_REPLY_ARRAY);
+        ASSERT(replies[i] != NULL && replies[i]->elements == 500);
     }
     t2 = usec();
     for (i = 0; i < num; i++) freeReplyObject(replies[i]);
@@ -660,7 +666,7 @@ static void test_throughput(struct config config) {
 //      * arrive in a single packet, causing all callbacks to be executed in
 //      * a single pass). */
 //     while(__test_callback_flags == 0) {
-//         assert(redisBufferRead(c) == REDIS_OK);
+//         ASSERT(redisBufferRead(c) == REDIS_OK);
 //         redisProcessCallbacks(c);
 //     }
 //     test_cond(__test_callback_flags == 0x010203);
-- 
1.9.1
```

**Comment by badboy** [comment]
@keith-bennett-gbg: just create a new Pull Request

**Comment by keith-bennett-gbg** [comment]
Yeah sorry about that. I haven't used git repos other than my own prior to this.

**Comment by badboy** [comment]
No problem, everything is fine, will read code in #309 tomorrow

**Comment by badboy** [comment]
Closed by 27076a3e6d60f3c98d116df5d6b7e1bfdd5d970b
